File: fe-admin/src/app/core/service/staff.service.ts
import {inject, Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {ENV} from "../../environment";
import {SuccessResponse} from "../model/response/success-response";
import {PageData} from "../model/response/page-data";
import {StaffResponse} from "../model/response/staff-response";
import {SearchStaffRequest} from "../model/request/search-staff-request";
import {CreateStaffRequest} from "../model/request/create-staff-request";
import {UpdateStaffRequest} from "../model/request/update-staff-request";

@Injectable({
    providedIn: 'root',
})
export class StaffService {
    private readonly url = ENV.API_BASE_URL + "clinic/staffs";
    private readonly http = inject(HttpClient);

    search(request: SearchStaffRequest): Observable<SuccessResponse<PageData<StaffResponse>>> {
        return this.http.post<SuccessResponse<PageData<StaffResponse>>>(`${this.url}/search`, request);
    }

    findById(id: number): Observable<SuccessResponse<StaffResponse>> {
        return this.http.get<SuccessResponse<StaffResponse>>(`${this.url}/${id}`);
    }

    create(request: CreateStaffRequest, file: File | null): Observable<SuccessResponse<StaffResponse>> {
        return this.http.post<SuccessResponse<StaffResponse>>(this.url, this.toFormData(request, file));
    }

    update(id: number, request: UpdateStaffRequest, file: File | null): Observable<SuccessResponse<StaffResponse>> {
        return this.http.put<SuccessResponse<StaffResponse>>(`${this.url}/${id}`, this.toFormData(request, file));
    }

    delete(id: number): Observable<SuccessResponse<void>> {
        return this.http.delete<SuccessResponse<void>>(`${this.url}/${id}`);
    }

    private toFormData(request: CreateStaffRequest | UpdateStaffRequest, file: File | null): FormData {
        const formData = new FormData();
        formData.append("request", new Blob([JSON.stringify(request)], {type: "application/json"}));
        if (file) {
            formData.append("file", file);
        }
        return formData;
    }
}

File: be-api/src/main/java/cd/beapi/service/StaffService.java
package cd.beapi.service;

import cd.beapi.dto.request.CreateStaffRequest;
import cd.beapi.dto.request.SearchStaffRequest;
import cd.beapi.dto.request.UpdateStaffRequest;
import cd.beapi.dto.request.UpdateStaffStatusRequest;
import cd.beapi.dto.response.PageData;
import cd.beapi.dto.response.PublicDentistResponse;
import cd.beapi.dto.response.StaffResponse;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface StaffService {
    PageData<StaffResponse> search(SearchStaffRequest searchStaffRequest);

    List<PublicDentistResponse> findDentistOptions();

    StaffResponse findCurrentDentist(String username);

    StaffResponse findById(Long id);

    StaffResponse save(CreateStaffRequest createStaffRequest, MultipartFile file);

    StaffResponse update(Long id, UpdateStaffRequest updateStaffRequest, MultipartFile file);

    StaffResponse updateStatus(Long id, UpdateStaffStatusRequest updateStaffStatusRequest);
}
